sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -167,36 +167,43 @@ object InMemoryFileIndex extends Logging {
       hadoopConf: Configuration,
       filter: PathFilter,
       sparkSession: SparkSession): Seq[(Path, Seq[FileStatus])] = {
+    // Filter out the directory before listing it leaf files
+    val filteredPaths = if (filter != null) {
+      paths.filter(filter.accept)
+    } else {
+      paths
+    }
 
     // Short-circuits parallel listing when serial listing is likely to be faster.
-    if (paths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
-      return paths.map { path =>
+    if (filteredPaths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
+      return filteredPaths.map { path =>
         (path, listLeafFiles(path, hadoopConf, filter, Some(sparkSession)))
```

Comment:
Rather than change it this way, should `listLeafFiles` just check if `filter` matches `path` initially?

Reply:
this maybe still need to trigger a spark job, because the check is pushed down into the executor side.